### PR TITLE
Humble ignition gazebo

### PR DESCRIPTION
**Pull request by mhubii**
### Ignition Gazebo
Ignition Gazebo (Gazebo) will be the predecessor of Gazebo (Gazebo Classic). This PR aims to already support Ignition Gazebo.

Currently, most functionality in this PR works, except for the interactive marker for MoveIt 2. For early adoption, this repo can be checked out `dev-humble-ignition-gazebo`. It can be run as usual.

[iiwa7.webm](https://github.com/lbr-stack/lbr_fri_ros2_stack/assets/26366414/d34ef5b7-3726-4d27-8d5d-6f0a00550916)

### Tasks
- [ ] Add Gazebo install hint 
  - https://gazebosim.org/docs/harmonic/ros_installation
  - https://gazebosim.org/docs/fortress/ros_installation#installing-the-default-gazebo-version-for-a-ros-distribution-using-binary-installations
- [ ] Update launch files
- [ ] Update Gazebo plugin in robot description
- [ ] Add conditional dependencies
  - [ ] `ros-humble-ros-gz`
  - [ ] `ros-humble-gazebo-ros`

### CODE REVIEW

File: lbr_bringup/package.xml [hunk 1]
```diff
@@ -20,6 +20,7 @@
   <exec_depend>robot_state_publisher</exec_depend>
   <exec_depend>gazebo_ros</exec_depend>
   <exec_depend>gazebo_ros2_control</exec_depend>
+  <exec_depend>ign_ros_control</exec_depend>
```
Comment (mhubii) [inline lbr_bringup/package.xml]: depends

File: lbr_description/gazebo/lbr.gazebo.xacro [hunk 2]
```diff
@@ -4,7 +4,7 @@
 
         <!-- ros_control-plugin -->
         <gazebo>
-            <plugin name="gazebo_ros2_control" filename="libgazebo_ros2_control.so">
+            <plugin name="ign_ros2_control::IgnitionROS2ControlPlugin" filename="ign_ros2_control-system">
```
Comment (mhubii) [inline lbr_description/gazebo/lbr.gazebo.xacro]: depends

File: lbr_description/lbr_description.dsv [hunk 3]
```diff
@@ -1 +1 @@
-prepend-non-duplicate;GAZEBO_MODEL_PATH;share
```
Comment (mhubii) [inline lbr_description/lbr_description.dsv:1]: depends

File: lbr_description/ros2_control/lbr.ros2_control.xacro [hunk 4]
```diff
@@ -17,7 +17,7 @@
             <!-- define hardware including parameters, also gazebo -->
             <xacro:if value="${sim}">
                 <hardware>
-                    <plugin>gazebo_ros2_control/GazeboSystem</plugin>
```
Comment (mhubii) [inline lbr_description/ros2_control/lbr.ros2_control.xacro]: depends

**Comment by mhubii** [comment]
blocked by https://github.com/ros-controls/gz_ros2_control/issues/182

File: lbr_description/lbr_description/launch_mixin.py [hunk 5]
```diff
@@ -42,15 +42,52 @@ def node_spawn_entity(
                 "robot_description",
                 "-entity",
                 LaunchConfiguration("robot_name"),
-                "-robot_namespace",
```
Comment (mhubii) [inline lbr_description/lbr_description/launch_mixin.py]: verify this functions

**Comment by mhubii** [comment]
interactive marker nonreactive for moveit launches

**Comment by mhubii** [comment]
read up on migration guide: https://gazebosim.org/docs/harmonic/migrating_gazebo_classic_ros2_packages

**Comment by jclinton830** [comment]
I tried to run this locally, and there is a dependency problem with the fri_vendor package.

**Comment by mhubii** [comment]
hi @jclinton830 and thank you for the feedback. This PR was left a little behind, hence why it likely doesn't work.

**Comment by jclinton830** [comment]
@mhubii Any idea on when this will be merged into the humble branch? Happy to assist with the process if it expedites it.

**Comment by mhubii** [comment]
hi @jclinton830 and sorry for the late response. Happy to review your PR if you decide to contribute!

Best way to do this is to start from the `humble` branch. These need to be implemented:

- Implement the launch for the new Gazebo
- Replace the old Gazebo plugin here:
https://github.com/lbr-stack/lbr_fri_ros2_stack/blob/71bf3f8957697b31fb887106c5dbe02074d9a9ea/lbr_ros2_control/config/lbr_system_interface.xacro#L18
- Replace the old Gazebo plugin here:
https://github.com/lbr-stack/lbr_fri_ros2_stack/blob/71bf3f8957697b31fb887106c5dbe02074d9a9ea/lbr_description/gazebo/lbr_gazebo.xacro#L7
- Update dependencies in `package.xml`

You can get inspiration:

- Here: https://github.com/ros-controls/gz_ros2_control/tree/humble
- And in this PR

**Comment by jclinton830** [comment]
Hi @mhubii I have forked your humble version and have made my own branch called dev-humble-ign. 

https://github.com/acfr/lbr_fri_ros2_stack/tree/dev-humble-ign

Do you take PRs directly into humble from forks?

**Comment by mhubii** [comment]
sorry for the late response @jclinton830. This branch should now work.

One remaining issue: Somehow the new Gazebo seems slow. The `sdf` (simulation description format), which is loaded here:

https://github.com/lbr-stack/lbr_fri_ros2_stack/blob/833ed94a5b773035864ba28479bad179a7867278/lbr_bringup/lbr_bringup/gazebo.py#L23

and accessed from:

`/usr/share/ignition/ignition-gazebo6/worlds/empty.sdf`

has an integrator step size (`max_step_size`):

```xml
    <physics name="1ms" type="ignored">
      <max_step_size>0.001</max_step_size>
```

If you set this from `0.001 -> 0.01`, then the simulation is quicker.

One might need to add an `sdf` file to this repo and extend the environment variable `GZ_SIM_RESOURCE_PATH`.

You can target PRs against this branch (`dev-humble-ignition-gazebo`)

**Comment by jclinton830** [comment]
> sorry for the late response @jclinton830. This branch should now work.
> 
> One remaining issue: Somehow the new Gazebo seems slow. The `sdf` (simulation description format), which is loaded here:
> 
> https://github.com/lbr-stack/lbr_fri_ros2_stack/blob/833ed94a5b773035864ba28479bad179a7867278/lbr_bringup/lbr_bringup/gazebo.py#L23
> 
> and accessed from:
> 
> `/usr/share/ignition/ignition-gazebo6/worlds/empty.sdf`
> 
> has an integrator step size (`max_step_size`):
> 
> ```
>     <physics name="1ms" type="ignored">
>       <max_step_size>0.001</max_step_size>
> ```
> 
> If you set this from `0.001 -> 0.01`, then the simulation is quicker.
> 
> One might need to add an `sdf` file to this repo and extend the environment variable `GZ_SIM_RESOURCE_PATH`.
> 
> You can target PRs against this branch (`dev-humble-ignition-gazebo`)

What is the average real-time factor you are witnessing? In the above branch that I made in my forked version gazebo is working fast, with an average real-time factor of 90%.
 
Also, have you fixed the problem of not being able to move the robot using the interactive marker?

**Comment by mhubii** [comment]
> What is the average real-time factor you are witnessing? In the above branch that I made in my forked version gazebo is working fast, with an average real-time factor of 90%.

With integrator step size `0.001` ~ `20%` and with `0.01` about ~ `100%`.



> Also, have you fixed the problem of not being able to move the robot using the interactive marker?

Yes that's working now. You can e.g. try:

```shell
ros2 launch lbr_bringup gazebo.launch.py \
    model:=med7
```

and

```shell
ros2 launch lbr_bringup move_group.launch.py \
    rviz:=true \
    mode:=gazebo \
    model:=med7
```

If you could add a way to load a custom `.sdf` file, that would be great! Feel free to target a PR against `dev-humble-ignition-gazebo`. I can merge and modify as needed.

Note you can do this even after this PR gets merged.

**Pull request merged**